### PR TITLE
Update NUMA nodes scoring KEP to consider zone distance when scoring

### PR DESCRIPTION
**Pull request by PiotrProkop**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Update  `LeastNUMANodes` ScoringStrategy  for `NodeResourceTopologyMatch` to take distance between topology zones into consideration.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by PiotrProkop** [comment]
@ffromani @swatisehgal please review

**Review by ffromani — COMMENTED** [review]
high level review: I'm fine with the idea and it's a nice feature. Questions about implemention:

1. when we talk about distances, are considering abstract values or do we prefer (need?) to use the actual value reported by the linux kernel? if so, do we need to expose these values to the scheduler?
2. (more like thinking aloud than actual question) we will need to figure out how to expose the flag `prefer-closest-numa-nodes`, this is something we are figuring out nowadays (see: https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/issues/24 and related)

**Comment by PiotrProkop** [comment]
> high level review: I'm fine with the idea and it's a nice feature. Questions about implemention:
> 
>     1. when we talk about distances, are considering abstract values or do we prefer (need?) to use the actual value reported by the linux kernel? if so, do we need to expose these values to the scheduler?
> 
>     2. (more like thinking aloud than actual question) we will need to figure out how to expose the flag `prefer-closest-numa-nodes`, this is something we are figuring out nowadays (see: [k8stopologyawareschedwg/noderesourcetopology-api#24](https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/issues/24) and related)

1. We already have this info in NRTs, it's called `Costs` and NFD reads it from `sysfs` https://github.com/kubernetes-sigs/node-feature-discovery/blob/1aa76029ac184b612001746932d125ff39373913/pkg/resourcemonitor/noderesourcesaggregator.go#L348
2. For now I assumed that it can be enabled by default and later when we will have a support for exposing `TopologyManagerPolicyOptions` we can make smarter decision.

**Comment by ffromani** [comment]
>     1. We already have this info in NRTs, it's called [Costs](https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/blob/master/pkg/apis/topology/v1alpha1/types.go#L56) and NFD reads it from `sysfs` [here](https://github.com/kubernetes-sigs/node-feature-discovery/blob/1aa76029ac184b612001746932d125ff39373913/pkg/resourcemonitor/noderesourcesaggregator.go#L348)

Great, now that you point this out, we implemented this ages ago and I'm glad this is eventually coming to fruition :)

>      2. For now I assumed that it can be enabled by default and later when we will have a support for exposing `TopologyManagerPolicyOptions` we can make smarter decision.

That would be fine for me.

**Comment by PiotrProkop** [comment]
> > ```
> > 1. We already have this info in NRTs, it's called [Costs](https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/blob/master/pkg/apis/topology/v1alpha1/types.go#L56) and NFD reads it from `sysfs` [here](https://github.com/kubernetes-sigs/node-feature-discovery/blob/1aa76029ac184b612001746932d125ff39373913/pkg/resourcemonitor/noderesourcesaggregator.go#L348)
> > ```
> 
> Great, now that you point this out, we implemented this ages ago and I'm glad this is eventually coming to fruition :)
> 
> > ```
> >  2. For now I assumed that it can be enabled by default and later when we will have a support for exposing `TopologyManagerPolicyOptions` we can make smarter decision.
> > ```
> 
> That would be fine for me.

For further reference I'll modify this KEP to include this points.
@ffromani can you /ok-to-test for me?

**Comment by ffromani** [comment]
> > > ```
> > > 1. We already have this info in NRTs, it's called [Costs](https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/blob/master/pkg/apis/topology/v1alpha1/types.go#L56) and NFD reads it from `sysfs` [here](https://github.com/kubernetes-sigs/node-feature-discovery/blob/1aa76029ac184b612001746932d125ff39373913/pkg/resourcemonitor/noderesourcesaggregator.go#L348)
> > > ```
> > 
> > 
> > Great, now that you point this out, we implemented this ages ago and I'm glad this is eventually coming to fruition :)
> > > ```
> > >  2. For now I assumed that it can be enabled by default and later when we will have a support for exposing `TopologyManagerPolicyOptions` we can make smarter decision.
> > > ```
> > 
> > 
> > That would be fine for me.
> 
> For further reference I'll modify this KEP to include those points. @ffromani can you /ok-to-test for me?

absolutely (just forgot to do this earlier)

**Comment by PiotrProkop** [comment]
KEP updated to include info about:
* how scheduler will retrieve distance(cost) information
* calculating average distance will be enabled by default until `TopologyManagerOptions` will be exposes by NRT API(or NFD)
*

**Comment by ffromani** [comment]
> KEP updated to include info about:
> 
>     * how scheduler will retrieve distance(cost) information
> 
>     * calculating average distance will be enabled by default until `TopologyManagerOptions` will be exposes by NRT API(or NFD)

thanks. Provisional ACK/LGTM about the general direction, I'll have proper review in the coming days

**Review by swatisehgal — COMMENTED** [review]
Nice to finally see us moving in a direction where we are taking NUMA distances into consideration when making scheduling decisions!

A big +1 for the feature!

Leaving a few comments. I wasn't able to review the whole KEP, will get back to it in the coming days.

File: kep/454-numa-nodes-scoring/README.md [hunk 1]
```diff
@@ -59,10 +63,14 @@ The narrower bitmask is preffered or if the same number of bits is set then the
 
 To calculate how many NUMA nodes are required to run given container/pod the plugin will:
 * generate possible combinations starting from narrowest possible bitmask to widest
+* find minimum average distance for combinations of given length
 * the [Combination][2] function can used to generate combinations since it generate combinations from the lowest values
 * combine resources available in combination by iterating over every Resources present in every NUMA node and adding them together
 * each NUMA nodes combination will be evaluated to see if there are enough resources to satisfy POD requirements
+* the average distances between nodes will be calculated for every combination, if it equals minimum average distance, return the combination
+  if the minimum average cannot be find, look for combination with lowest average within given mask length
```
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply kep/454-numa-nodes-scoring/README.md]: 👍

File: kep/454-numa-nodes-scoring/README.md [hunk 2]
```diff
@@ -74,28 +82,42 @@ Calculating required NUMA nodes for a `pod` scope is straightforward, we just ne
 When it comes to `container` scope we need to calculate `required` NUMA nodes for every container in proper order([same as TopologyManager asks for hints](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/scope_container.go#L52))
```
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply kep/454-numa-nodes-scoring/README.md]: 👍

**Comment by PiotrProkop** [comment]
@swatisehgal thanks for the review! I have applied all your suggestions.

File: kep/454-numa-nodes-scoring/README.md [hunk 3]
```diff
@@ -47,6 +47,14 @@ to get available resources on the worker nodes and identify which topology polic
 For now available ScoringStrategies are running only for `single-numa-node` TopologyManager policy.
 `LeastNUMANodes` strategy can score nodes running all TopologyManager policies.
 
+Kubernetes `v1.26` introduced `TopologyManagerPolicyOptions` that allows to control `TopologyManager` behaviour.
+New option `prefer-closest-numa-nodes` will favor sets of NUMA nodes with shorter distance between them when making admission decisions.
+`LeastNUMANodes` ScoringStrategy will also take distance between topology zones into consideration and favor nodes with the smallest average distance between zones.
+Distances between topology zones are exposed as `Costs` field in [NRT CRD](https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/blob/master/pkg/apis/topology/v1alpha1/types.go#L56).
+This behaviour will be enabled by default, since [CRD][1] doesn't expose `TopologyManagerPolicyOptions`,
```
Comment (ffromani) [inline kep/454-numa-nodes-scoring/README.md:54]: nit: ...since the CRD...

**Review by ffromani — COMMENTED** [review]
/lgtm
/hold
giving time to @swatisehgal to do another pass

File: kep/454-numa-nodes-scoring/README.md (hunk 3 above)
Comment (PiotrProkop) [reply kep/454-numa-nodes-scoring/README.md]: fixed

Comment (ffromani) [reply kep/454-numa-nodes-scoring/README.md]: Thanks! for the future, no need from my side to re-upload *just* for so tiny NITs.

**Comment by ffromani** [comment]
/lgtm
(same as https://github.com/kubernetes-sigs/scheduler-plugins/pull/484#pullrequestreview-1271047236)

**Review by swatisehgal — APPROVED** [review]
/lgtm
/unhold

/assign @Huang-Wei 
for approval

File: kep/454-numa-nodes-scoring/README.md [hunk 4]
```diff
@@ -47,6 +47,14 @@ to get available resources on the worker nodes and identify which topology polic
 For now available ScoringStrategies are running only for `single-numa-node` TopologyManager policy.
 `LeastNUMANodes` strategy can score nodes running all TopologyManager policies.
 
+Kubernetes `v1.26` introduced `TopologyManagerPolicyOptions` that allows to control `TopologyManager` behaviour.
```
Comment (Huang-Wei) [inline kep/454-numa-nodes-scoring/README.md:50]: Just a nit: we can merge the KEP for now. However, in terms of impl., if this depends on 1.26, or can only get the desired effect in 1.26, we should wait for `master` in scheduler-plugins to 1.26, for now it's still 1.25.

**Review by Huang-Wei — COMMENTED** [review]
/approve

File: kep/454-numa-nodes-scoring/README.md (hunk 4 above)
Comment (PiotrProkop) [reply kep/454-numa-nodes-scoring/README.md:50]: Makes sense, @Huang-Wei  are there any plans on supporting 1.26 or maybe I can help with 1.26 support?

Comment (Huang-Wei) [reply kep/454-numa-nodes-scoring/README.md:50]: commented in #489.

**Pull request merged**